### ISSUE TITLE
malheur 

### ISSUE THREAD

**Issue by SeanKim777** [spender-sandbox/cuckoo-modified#335]
Hi
Issue #193 seems created by malheur.py reporting module [Check source here](https://github.com/spender-sandbox/cuckoo-modified/blob/master/modules/reporting/malheur.py#L170)

I have found generating melheur report take too long.
14282 malheur dataset( == number of analysis) take more than 2mins. And it is expected to take more for more dataset.

`$ ls -l ./test_dataset/ |wc -l`
`14282`
`$ time ./malheur -c ~/cuckoo/conf/malheur.conf -o ./test_increment.txt cluster ./test_dataset/`
`real	2m6.502s`
`user	6m27.744s`
`sys	0m1.729s`

malheur seems using multi-CPU while it's executing by looking at CPU usage and above user time.
It may hard to improving performance by updating malheur. (tested with malheur downloaded and compiled from [malheur git repo](https://github.com/rieck/malheur.git)  on today 25 Oct 2016)

So if there are more than 10K analyses then malheur reporint module need to run on a regular basis rather than each analysis like retention module.

Any suggestion?

**Comment by mallorybobalice** [comment]
https://github.com/rieck/malheur/commits/master

Oct commit is not meaningful

I've been running the one prior to it  on 30k /week rolling file sets for a while and yes,  mh takes longer that processing and reporting heh(hell, most of what my server seems to bedoing is freezing on static pdf until timeout and malheur 

>malheur report module need to run on a regular basis rather than each analysis like retention module.

We need to talk to @spender-sandbox and @rieck 

I'm unclear if we run it in incremental mode or not, and if we can. See https://github.com/rieck/malheur/issues/12

I'm unfamiliar with mh internals so if it parses all 30000 reports or md5 \txt files every file that probably explains why

**Comment by mallorybobalice** [comment]
Also whether it classifies using large reports, or signatures or summary reports

**Comment by rieck** [comment]
Hi all,

it's cool to see that you are looking at Malheur! 

I am not sure what you are actually trying to achieve. Clustering 14k behavior reports in about 2 minutes is not that bad. When we implemented the tool in 2010 similar experiments took much longer (see page 22 here https://www.tu-braunschweig.de/Medien-DB/sec/pubs/2011-jcs.pdf)

The tool has been designed to run in an incremental mode, where you would take the catch of the day (freshly analyzed reports) and pass it to the tool for finding similarities with old samples as well as identify novel clusters. Depending on the amount of data, this will take some time. In comparison to the actual sandbox analysis time, however, this should be marginal. 

The main problem I see with Malheur is long-time deployment. Over the time the collected clusters will grow and thus the tool will slow down. We have never explored this issue due to a lack of a stable data source (someone running thousands of malware samples through a sandbox on a regular basis). 

Regards, Konrad

**Comment by SeanKim777** [comment]
Thanks for information and comment @mallorybobalice and @rieck 
@rieck the reason why I have made this issue on cuckoo repo instead of malheur is 2 minutes processing time for 14K reports could be regarded as short. but if cuckoo need to processing thousands of suspicious samples in a short period of time, malheur processing part could become a bottleneck of entire process. and this is what I am currently having.

I have limited understanding and knowledge about malheur.
So, Please correct me if I made any wrong approach to solve this issue.

I have not fully understand issue [rieck/malheur#12](https://github.com/rieck/malheur/issues/12) created by @spender-sandbox because I have got different result on incremental execution on my datasets. I have used man page example syntax and divided datasets into 3 chunks.
`malheur -c cuckoocfg -s /tmp/state -o out1.txt -v --reset increment dataset1.zip`
`malheur -c cuckoocfg -s /tmp/state -o out2.txt -v increment dataset2.zip`
`malheur -c cuckoocfg -s /tmp/state -o out3.txt -v increment dataset2.zip`

I have got similar result created by cuckoo malheur reporting module [source](https://github.com/spender-sandbox/cuckoo-modified/blob/master/modules/reporting/malheur.py#L170)
Total execution time is almost similar between incremental and cluster mode.
But as incremental will take less time for a chunk. I think executing malheur in a incremental mode could lower cuckoo report generating time. (means changing malheur.py to execute malheur in incremental mode and merge output into one result file)

@rieck according to your answer above, can I use this approach for the every new reports?

another question: '-s' option. should I keep 'prototypes.zfa' and 'rejected.zfa' to get a same result for the incremental mode?

**Comment by rieck** [comment]
Okay, let me comment on these issues:

- Maybe I a missing something. Generating 14k reports with Cuckoo Sandbox takes some time. For example, if we have 100 machines and each runs a sample for only 1 minute, we need at least 2 hours. Is the remaining 2 minutes needed by Malheur really an issue?

- Malheur has not be designed to process individual reports. Note you cannot cluster a single report but only assign it to an existing cluster or wait until more data is available. Thus, it is better to collect reports and pass them to the tool in larger chunks.

- Malheur will not merge any output logs, if I recall correctly. However, it will generate a global state at the location provided by `-s`. 

- Clustering methods are inherently quadratic it run-time. In a practical deployment you would just prefer incremental mode, as it only requires this quadratic run-time for the new data.

Hope this helps,
Konrad

**Comment by mallorybobalice** [comment]
Alight, so basically I have the same problem. 
I run process

**Comment by mallorybobalice** [comment]
Ps edited plz check back

**Comment by rieck** [comment]
I would recommend a larger chunk size, say 100 or even 500. Malheur implements a more or less complex feature extraction process. Please refer to the journal paper for details: https://www.sec.cs.tu-bs.de/pubs/2011-jcs.pdf

**Comment by SeanKim777** [comment]
@mallorybobalice and @rieck 
Thanks for comment and kind information.
I'm working on update malheur report module as advised
will post it after I finished testing. not sure how long it take though

**Comment by mallorybobalice** [comment]
@spender-sandbox  final bump :( ?

**Comment by spender-sandbox** [comment]
We don't bump stuff here -- I see all the messages fine.  The problem remains that for many/most people's usage of Malheur, when they submit a sample they'd like to be able to immediately see what similar samples exist.  The lock necessitates switching to a mode where that can't happen because it needs to be performed in chunks.  I don't know myself how to resolve that, if I did I would comment.  Or if you do, comment about that, but please don't just bump.

-Brad

**Comment by rieck** [comment]
Potentially off-topic: Would it make sense to have a tool that simply computes the similarity of one report to a group of previous reports and return the top most similar reports? This could be done much quicker than clustering --- definitely $O(n)$ but likely even faster.

**Comment by spender-sandbox** [comment]
Yes, that's pretty much what we were going for with the "similar" tab -- we ended up cutting off results actually because in some instances there are too many similar samples.

-Brad

**Comment by rieck** [comment]
Interesting, we have new research project that deals with analysis of malware behaviour. We are using neither Cuckoo nor Malheur. Nonetheless, something like a simple "similarity checker" could be done as part of this project.

@chwress: What do you think?

**Comment by mallorybobalice** [comment]
@spender-sandbox, 
Let me elaborate. Not bump bump, but
>when they submit a sample they'd like to be able to immediately see what similar samples exist.

I get this, like I said though for my case… at say 30k rolling mh reports and txt files, areat 5m=16'

**Comment by mallorybobalice** [comment]
@rieck did you mean one of your recent projects in gh? I don't necesarilly mean for cuckoo modified but just in principle

**Comment by rieck** [comment]
I was talking about a "real" project. See here http://www.vamos-project.org

;)
